pymbolic/primitives.py

```diff
@@ -20,19 +23,42 @@
 THE SOFTWARE.
 """
 
-from abc import ABC, abstractmethod
+from dataclasses import dataclass, fields
 from sys import intern
-from typing import ClassVar, Dict
+from typing import Any, Callable, ClassVar, Mapping, Tuple, Union
+from warnings import warn
 
 import pymbolic.traits as traits
 
 
+# FIXME: This is a lie. Many more constant types (e.g. numpy and such)
+# are in practical use and completely fine. We cannot really add in numpy
+# as a special case (because pymbolic doesn't have a hard numpy dependency),
+# and there isn't a usable numerical tower that we could rely on. As such,
+# code abusing what constants are allowable will have to type-ignore those
+# statements. Better ideas would be most welcome.
+#
+# References:
+# https://github.com/python/mypy/issues/3186
+# https://discuss.python.org/t/numeric-generics-where-do-we-go-from-pep-3141-and-present-day-mypy/17155/14
+_ConstantT = Union[int, float, complex]
```

Comment:
Could we do smth like:

```suggestion
try:
    import numpy as np
    _ConstantT = Union[int, float, complex, np.generic]
except ImportError:
    _ConstantT = Union[int, float, complex]
```
?

Reply:
Take a look at the latest revision, see how you like it.

Reply:
This looks good, thanks!